### ISSUE TITLE
The game freezes on Save & Exit

### ISSUE THREAD

**Issue by HSteinway** [big-ab-games/robo-instructus#236]
Every time. =\ Have to close it through the Task Manager.

Win7 x64, 8 gb, GTX 1660

**Comment by alexheretic** [comment]
Thanks for reporting. Do you have anything in your error.log in the game folder?

**Comment by HSteinway** [comment]
> Thanks for reporting. Do you have anything in your error.log in the game folder?

Apparently I don't have this file. Is has to be in the SteamLibrary/steamapps/common/robo-instructus folder? There's no such file. =\

**Comment by alexheretic** [comment]
Yep that's where it should be. If it's hanging it may not produce an error there I suppose.

I can't reproduce this on my Windows setup (which is win10). I assume this didn't occur with the previous version you used.

Can you try the last few versions and see if it was introduced by one? I'll upload some historic demo builds to https://github.com/big-ab-games/robo-instructus/releases/tag/1.28

**Comment by HSteinway** [comment]
All demo builds are working nice. I'll try to reinstall the game.

Upd.: Didn't help.

**Comment by alexheretic** [comment]
That's strange, but I suppose good news.

The demo builds are quite close to the full builds really. They don't have
steam integration, but I wouldn't expect that to be an issue.

**Comment by alexheretic** [comment]
If you go to the title screen and create another clean profile, does that behave the same way?

**Comment by HSteinway** [comment]
> If you go to the title screen and create another clean profile, does that behave the same way?

I have the second profile, that just started a while ago, it freezes too, the brand new one works fine just until the moment I write a line. Then it freezes too.

**Comment by alexheretic** [comment]
So to be clear, when you start a new clean profile you can save & exit
normally, until you start actually coding after which it hangs exiting?

Could be a problem with writing the saves maybe. However, the demo builds
write saves in exactly the same way. Hmm.

Tomorrow I'll make a build with a bunch of debug logging and push it to a
steam branch. This should narrow down where is hanging.

**Comment by HSteinway** [comment]
> So to be clear, when you start a new clean profile you can save & exit
normally, until you start actually coding after which it hangs exiting?

Yep. Btw I didn't try to code in demo builds. On it now.

Upd.: I'm stupid, they share saves. Demo builds do not hang.

**Comment by alexheretic** [comment]
Yep they both involve writing to the same files. So it probably isn't that. One difference is steam integration.

Can you try running the main game without steam? You do this by quitting steam so that it isn't running and running the game exe from the folder directly. If steam isn't detected it'll just run without the steam features.

**Comment by HSteinway** [comment]
> Can you try running the main game without steam? You do this by quitting steam so that it isn't running and running the game exe from the folder directly. If steam isn't detected it'll just run without the steam features.

It works fine without Steam! Saves and exits.

**Comment by alexheretic** [comment]
Ok thanks for your patience. So it seems it's an issue with shutting down the steam client.

I'll build and deploy a test branch with
* Logging code to confirm where the game hangs.
* Optional arg `--nosteam` to run without steam, even if steam itself _is_ running.
* Another build with the steam sdk updated from 1.46 -> 1.48a.

Give me a sec...

**Comment by alexheretic** [comment]
Ok if you right click on the game and go to _BETAS_ you can select **steam-1.48a-logging**. If we're lucky the steam-sdk update will help.

Try to load the game up & exit. Then look at your _error.log_, for example I get

```
2020-05-27T08:16:24Z (main) control_flow = Exit
2020-05-27T08:16:24Z (compute) Done final save
2020-05-27T08:16:24Z (compute) Dropping steam client...
2020-05-27T08:16:24Z (compute) Dropped steam client
2020-05-27T08:16:25Z (main) Render finished
2020-05-27T08:16:25Z (main) Compute finished
```

If it still fails you can try setting your launch options to `--nosteam` which should allow you to run the game from steam without the steam features.

**Comment by HSteinway** [comment]
It crashes and says:
2020-05-27T08:22:59Z (main) control_flow = Exit
2020-05-27T08:22:59Z (compute) Done final save
2020-05-27T08:22:59Z (compute) Dropping steam client...
2020-05-27T08:22:59Z (compute) Dropped steam client

**Comment by alexheretic** [comment]
Ok that looks more like the render loop isn't exitting properly... which should have nothing to do with steam.

Does it work with `--nosteam` in your launch options, you can set that by right clicking on the game _GENERAL_ tab.

**Comment by HSteinway** [comment]
Hey, I'm sorry, got to sleep for a while).
--nosteam doesn't help

**Comment by alexheretic** [comment]
Ok it could instead be something to do with launching the game from steam,
rather than the steam integration itself. This is quite strange.

Can you try to add a demo build as a "non steam game" from the bottom left
menu in steam. If you launch it from steam does it start to hang too?

Another thing to try is disabling the steam overlay and seeing it that
helps.

**Comment by HSteinway** [comment]
Adding demo build 1.26: works.
Adding demo build 1.27: works.
Adding demo build 1.28: hangs.

Overlay is always off.

**Comment by alexheretic** [comment]
> Adding demo build 1.26: works.
Adding demo build 1.27: works.
Adding demo build 1.28: hangs.

Great thanks, so it looks like this was introduced in winit `0.22.2` as 1.27 used winit `0.22.1`.

**Comment by alexheretic** [comment]
I'll make a build simply downgrading winit which I think will address this for you. Give me a sec...

**Comment by alexheretic** [comment]
I've added a `winit-0.22.1` branch identical to the 1.28 release except the winit version. @HSteinway Can you please confirm that I'm on the right lines this time & this one does not hang?

**Comment by HSteinway** [comment]
> I've added a `winit-0.22.1` branch identical to the 1.28 release except the winit version. @HSteinway Can you please confirm that I'm on the right lines this time & this one does not hang?

IT WORKS! :-D 
![image](https://user-images.githubusercontent.com/58853753/83078109-7311e780-a081-11ea-9781-64545bb65f33.png)

**Comment by alexheretic** [comment]
Great we found it, thanks again for your patience!

I'll leave the change in a branch for now and see if the upstream issue gets anywhere. It may be difficult for the winit devs to reproduce. If I get time I'll see if I can get a Win7 running somewhere and try to reproduce myself.

**Comment by alexheretic** [comment]
I can't setup a windows 7 vm or bare metal install with my current hardware. So while we tracked this issue down I can't check future winit versions have fixed the issue and so can't guarantee that future Robo Instructus versions won't regress here. This isn't great for me supporting windows 7.

What I have done is released the winit-0.22.1 version as **1.28.1**.  I've also created a `win7` steam branch that can stick around forever.

I'm going to drop full support for windows 7 & 8 in future versions because I can't reproduce issues that are unique to them. However, I'll try to do my best to support the `win7` branch if I can, ie from indirect debugging like this.
